price-feeder/oracle/provider/huobi.go

```diff
@@ -27,140 +32,248 @@ type (
+		reconnectTicker *time.Ticker
+	}
+
+	// HuobiTicker defines the response type for the channel and
+	// the tick object for a given ticker/symbol
+	HuobiTicker struct {
+		CH   string    `json:"ch"` // Data belonged channel，Format：market.$symbol.ticker
+		Tick HuobiTick `json:"tick"`
+	}
+
+	// HuobiTick defines the response type for the last 24h market summary
+	// and the last traded price for a given ticker/symbol
+	HuobiTick struct {
+		Vol       float64 `json:"vol"`       // Accumulated trading value of last 24 hours
+		LastPrice float64 `json:"lastPrice"` // Last traded price
+	}
+
+	// HuobiSubscriptionMsg Msg to subscribe to one ticker channel at time
+	HuobiSubscriptionMsg struct {
+		Sub string `json:"sub"` // channel to subscribe market.$symbol.ticker
 	}
 )
 
-func NewHuobiProvider() *HuobiProvider {
-	return &HuobiProvider{
-		baseURL: huobiBaseURL,
-		client:  newDefaultHTTPClient(),
+func NewHuobiProvider(ctx context.Context, logger zerolog.Logger, pairs ...types.CurrencyPair) (*HuobiProvider, error) {
+	wsURL := url.URL{
+		Scheme: "wss",
+		Host:   huobiHost,
+		Path:   huobiPath,
 	}
-}
 
-func NewHuobiProviderWithTimeout(timeout time.Duration) *HuobiProvider {
-	return &HuobiProvider{
-		baseURL: huobiBaseURL,
-		client:  newHTTPClientWithTimeout(timeout),
+	wsConn, _, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to Huobi websocket: %w", err)
 	}
+
+	provider := &HuobiProvider{
+		wsURL:           wsURL,
+		wsClient:        wsConn,
+		logger:          logger.With().Str("module", "oracle").Logger(),
+		tickers:         map[string]HuobiTicker{},
+		reconnectTicker: time.NewTicker(huobiReconnectTime),
+	}
+
+	if err := provider.subscribeTickers(pairs...); err != nil {
+		return nil, err
+	}
+	provider.subscribedPairs = pairs
+
+	go provider.handleWebSocketMsgs(ctx)
+
+	return provider, nil
 }
 
-func (p HuobiProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
-	tickerPrices := make(map[string]TickerPrice, len(pairs))
-	for _, cp := range pairs {
-		price, err := p.getTickerPrice(cp.String())
-		if err != nil {
-			return nil, err
+// subscribeTickers subscribe to all currency pairs.
+func (p *HuobiProvider) subscribeTickers(cps ...types.CurrencyPair) error {
+	for _, cp := range cps {
+		huobiSubscriptionMsg := newHuobiSubscriptionMsg(cp)
+		if err := p.wsClient.WriteJSON(huobiSubscriptionMsg); err != nil {
+			return err
 		}
-
-		tickerPrices[cp.String()] = price
 	}
 
-	return tickerPrices, nil
+	return nil
 }
 
-func (p HuobiProvider) getTickerPrice(ticker string) (TickerPrice, error) {
-	path := fmt.Sprintf("%s%s?symbol=%s", p.baseURL, huobiTradeEndpoint, strings.ToLower(ticker))
+func (p *HuobiProvider) handleWebSocketMsgs(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewMessage):
+			// time after to avoid asking for prices too frequently
```

Comment:
What does this comment mean?

Reply:
It was kind of to give a reason to why have the `time.After` 
Removed at https://github.com/umee-network/umee/pull/569/commits/9269b0c910897b95d3ce3a279368d7e45b31339a